bundles/auth/org.eclipse.smarthome.auth.oauth2client/src/main/java/org/eclipse/smarthome/auth/oauth2client/internal/OAuthFactoryImpl.java

```diff
@@ -0,0 +1,155 @@
+import org.osgi.service.component.annotations.Deactivate;
+import org.osgi.service.component.annotations.Reference;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Implementation of {@link OAuthManager}
+ *
+ * @author Michael Bock - Initial contribution
+ * @author Gary Tse - ESH adaptation
+ */
+@NonNullByDefault
+@Component
+public class OAuthFactoryImpl implements OAuthFactory {
+
+    private final Logger logger = LoggerFactory.getLogger(OAuthFactoryImpl.class);
+
+    @Nullable
+    private OAuthStoreHandler oAuthStoreHandler;
+    @Nullable
+    private HttpClientFactory httpClientFactory;
+
+    int tokenExpiresInBuffer = OAuthClientServiceImpl.DEFAULT_TOKEN_EXPIRES_IN_BUFFER_SECOND;
+
+    private final Map<String, OAuthClientServiceImpl> oauthClientServiceCache = new ConcurrentHashMap<>();
+
+    @Deactivate
+    public void deactivate() {
+        // close each service
+        for (OAuthClientServiceImpl clientServiceImpl : oauthClientServiceCache.values()) {
+            clientServiceImpl.close();
+        }
+        oauthClientServiceCache.clear();
+    }
+
+    @Override
+    public String createOAuthClientService(String tokenUrl, @Nullable String authorizationUrl, String clientId,
+            @Nullable String clientSecret, @Nullable String scope, @Nullable Boolean supportsBasicAuth) {
+
+        String uuidHandle = UUID.randomUUID().toString();
+
+        @SuppressWarnings("null")
+        OAuthClientServiceImpl clientImpl = OAuthClientServiceImpl.getInstance(uuidHandle, oAuthStoreHandler, tokenUrl,
+                authorizationUrl, clientId, clientSecret, scope, supportsBasicAuth, tokenExpiresInBuffer,
+                httpClientFactory);
+
+        oauthClientServiceCache.put(uuidHandle, clientImpl);
+        return uuidHandle;
+    }
+
+    @Override
+    public OAuthClientService getOAuthClientService(String handle) {
+        OAuthClientServiceImpl clientImpl = null;
+        try {
+            clientImpl = oauthClientServiceCache.get(handle);
+            if (clientImpl == null || clientImpl.isClosed()) {
+                // This happens after reboot, or client was closed without factory knowing; create a new client
+                // the store has the handle/config data
+                clientImpl = OAuthClientServiceImpl.getInstance(handle, oAuthStoreHandler, tokenExpiresInBuffer,
+                        httpClientFactory);
+                oauthClientServiceCache.put(handle, clientImpl);
+            }
+        } catch (OAuthException e) {
+            logger.error("Cannot get OAuth client service by handle {}", handle, e);
+        }
+        return clientImpl;
+    }
+
+    @SuppressWarnings("null")
+    @Override
+    public void ungetOAuthService(String handle) {
+        OAuthClientServiceImpl clientImpl = oauthClientServiceCache.get(handle);
+        if (clientImpl == null) {
+            logger.error("{} handle not found.  Cannot unregisterOAuthServie", handle);
```

Comment:
Same here: Either reduce to DEBUG or throw an exception.

Reply:
Changed to debug.  You're right this is not severe.